src/librustc_typeck/collect.rs

```diff
@@ -1532,6 +1504,11 @@ fn type_scheme_of_item<'a,'tcx>(ccx: &CrateCtxt<'a,'tcx>,
                                 it: &hir::Item)
                                 -> ty::TypeScheme<'tcx>
 {
+    // Computing the type scheme of an item is a discrete task:
+    let item_def_id = ccx.tcx.map.local_def_id(it.id);
+    let _task = ccx.tcx.dep_graph.in_task(DepNode::TypeScheme(item_def_id));
+    ccx.tcx.dep_graph.read(DepNode::Hir(item_def_id)); // we have access to `it`
```

Comment:
@michaelwoerister 

> This could be moved to compute_type_scheme_of_item() below, since we don't need a dependency from TypeScheme to Tcache, only the other way round (i.e. creating the TypeScheme does not really access the TCache). Right now we get an unnecessary circular dependency.

Yes, true. I debated about doing that. The circular dependency is harmless, but I guess it would indeed be better to push the new task creation down, if for no other reason than other hypothetical callers of `compute_type_scheme_of_item` (none exist at present), would get the benefit of a new task.

This is actually one of the more subtle points of this work. Introducing a new node is a useful optimization for the dependency graph when you have lazy computation like this. If you don't do it, you wind up with artificial edges -- e.g., imagine both `foo` and `bar` reference `baz`, and we check `foo` first. Then we will add an edge that `foo` writes to `TYPE(baz)`. When we check `bar` (and `baz`), we will get a read of `TYPE(baz)` only, because it is already computed. This means that if `foo` must be recomputed, we will also recompute both `bar` and `baz`, even though that may not be necessary.